feathr_project/feathr/source.py

```diff
@@ -46,17 +49,19 @@ class HdfsSource(Source):
 
         Args:
             name (str): name of the source
-            path (str): The location of the source data. 
-            event_timestamp_column (Optional[str]): The timestamp field of your record. As sliding window aggregation feature assume each record in the source data should have a timestamp column. 
-            timestamp_format (Optional[str], optional): The format of the timestamp field. Defaults to "epoch". Possible values are: 
+            path (str): The location of the source data.
+            preprocessing (Optional[Callable]): A preprocessing python function that transforms the source data for further feature transformation.
+            event_timestamp_column (Optional[str]): The timestamp field of your record. As sliding window aggregation feature assume each record in the source data should have a timestamp column.
+            timestamp_format (Optional[str], optional): The format of the timestamp field. Defaults to "epoch". Possible values are:
             - `epoch` (seconds since epoch), for example `1647737463`
             - `epoch_millis` (milliseconds since epoch), for example `1647737517761`
-            - Any date formats supported by [SimpleDateFormat](https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html). 
-                
+            - Any date formats supported by [SimpleDateFormat](https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html).
+
         """
-    def __init__(self, name: str, path: str, event_timestamp_column: Optional[str]= None, timestamp_format: Optional[str] = "epoch") -> None:
+    def __init__(self, name: str, path: str, preprocessing: Optional[Callable] = None, event_timestamp_column: Optional[str]= None, timestamp_format: Optional[str] = "epoch") -> None:
```

Comment:
We probably shouldn't allow pyudf on context source, as users may mess up the observation data and have a hard time understanding why.